src/cargo/util/toml.rs

```diff
@@ -1394,6 +1383,53 @@ fn normalize(package_root: &Path,
     ret
 }
 
+fn inferred_bin_path(bin: &TomlBinTarget, package_root: &Path, lib: bool, bin_len: usize) -> PathBuf {
+    // we have a lib with multiple bins, so the bins are expected to be located inside src/bin
+    if lib && bin_len > 1 {
+        return Path::new("src").join("bin").join(&format!("{}.rs", bin.name())).to_path_buf()
+    }
+
+    // we have a lib with one bin, so it's either src/main.rs, src/bin/foo.rs or src/bin/main.rs
+    if lib && bin_len == 1 {
+        let path = Path::new("src").join(&format!("main.rs"));
+        if package_root.join(&path).exists() {
+            return path.to_path_buf()
+        }
+
+        let path = Path::new("src").join("bin").join(&format!("{}.rs", bin.name()));
+        if package_root.join(&path).exists() {
+            return path.to_path_buf()
+        }
+
+        return Path::new("src").join("bin").join(&format!("main.rs")).to_path_buf()
+    }
+
+    // here we have a single bin, so it may be located in src/main.rs, src/foo.rs,
+    // srb/bin/foo.rs or src/bin/main.rs
+    if bin_len == 1 {
+        let path = Path::new("src").join(&format!("main.rs"));
+        if package_root.join(&path).exists() {
+            return path.to_path_buf()
+        }
+
+        let path = Path::new("src").join(&format!("{}.rs", bin.name()));
+        if package_root.join(&path).exists() {
+            return path.to_path_buf()
+        }
+
+        let path = Path::new("src").join("bin").join(&format!("{}.rs", bin.name()));
+        if package_root.join(&path).exists() {
+            return path.to_path_buf()
+        }
+
+        return Path::new("src").join("bin").join(&format!("main.rs")).to_path_buf()
+
+    }
+
+    // here we have multiple bins, so they are expected to be located inside src/bin
+    Path::new("src").join("bin").join(&format!("{}.rs", bin.name())).to_path_buf()
```

Comment:
@alexcrichton not sure about this, maybe we should support multiple bin targets (for example: `foo` and `bar`) in `src/foo.rs` and `src.bar.rs`... but what should we do with `src/main.rs`?